design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+
+ For ksqlDB to be a successful project we need to provide an awesome out of the box developer experience
+ and it should be super easy and fun to write powerful event streaming applications.
+ Currently this is somewhat diffificult to do, and it requires the use of multiple clients and multiple moving parts in order to build
+ a simple event sourcing / CQRS style application.
+ We should provide a new ksqlDB client that provides that currently missing single interaction point with the event streaming
+ platform so that we can improve the application development process.
+           
+## Motivation and background
+
+In order to increase adoption of ksqlDB we need the developer experience to be as slick as possible. Currently, in order to write
+a simple event streaming application with ksqlDB, up to 4 clients may be needed: 
+
+1. An HTTP client to access the current HTTP/REST API
+2. A WebSocket client to access the streaming query WebSockets endpoint
+3. A JDBC client to access data in JDBC tables that are synced with KSQL via connect, as the current support for
+pull queries does not make the data otherwise easily accessible.
+4. A Kafka client for producing and consuming messages from topics.
+
+This is a lot for an application to handle and creates a steep learning curve for application developers.
+
+Moreover, our current HTTP API is lacking, especially in the area of streaming query results which makes it difficult to write an application
+that uses this API.
+
+This KLIP proposes that we:
+
+* Create a ksqlDB client (initially in Java and JavaScript) as the primary interaction point to the event streaming platform
+for an application
+* Create a new server side implementation that supports the client and enables the functionality of the client to be delivered in a straightforward
+and efficient way
+* The creation of a new wire protocol for the client<->server communication.
+* Migrate the functionality of the current HTTP/REST API over to the new server implementation and retire the parts that are no longer needed.
+
+## What is in scope
+
+### The client
+
+* We will create a new client, initially in Java (and potentially in JavaScript). Clients for other languages such as Python and Go will follow later.
+* The client will initially support execution and streaming of queries (both push and pull), inserting of rows into streams, DDL operations and admin operations
+such as list and describe.
+* The client will support a reactive / async interaction style using Java CompletableFuture and Reactive Streams / JDK9 Flow API (or a copy thereof if we cannot upgrade from Java 8)
+* The client will also support a direct / synchronous interaction style.
+
+### The protocol
+
+We will create a new super simple streaming protocol designed to efficiently stream data from server to client or client to server.
+HTTP/REST is a poor fit for streaming data as it is designed for a request/response interaction style. Attempting to implement "streaming"
```

Comment:
What do you think about http/2? For e.g gRPC bi directional streaming is all on top of http/2

Reply:
To add more, we can also look into gRPC as a substrate for RPC as well? It has language bindings in so many languages and we should be able to wrap gRPC and get our clients done in many languages quickly?

Reply:
Hey @vinothchandar 

> To add more, we can also look into gRPC as a substrate for RPC as well? It has language bindings in so many languages and we should be able to wrap gRPC and get our clients done in many languages quickly?

I looked into gRPC (and Thrift), as similar reasoning applies as to why I don't think they would be a good solution.

* The generated clients have pretty ugly APIs and we'd need to wrap them in custom code to give them a nice simple idiomatic reactive streams interface. This negates much of the advantage of generation in the first place.
* We'd also have to adapt the flow control behaviour to something reactive streams friendly - again, fiddly work that we'd have to do on both client and server side.
* We'd have no control of the threading model (esp. on the server side) - this means it couldn't use Vert.x and integrate nicely with the rest of the API using Vert.x
* Sadly HTTP2 streaming wasn't designed to present a streaming API to the client. The streaming is really as a back channel for keeping HTTP caches in sync. I.e. HTTP2 can pre-emptively stream resources from a server to a browser cache so that when the browser requests a resource it's already cached locally. But that's an internal implementation detail of the browser and there's no API exposed to user JavaScript code to actually get streaming results - from the user's point of view it's just the same old request/response HTTP. This means you can't use HTTP2 as a good means of streaming to JavaScript browser clients, and this is the reason WebSockets haven't been deprecated :)

So to summarise: If we did use gRPC generation there'd be a lot of work adapting the generated clients to a useful form, plus we'd lose control of the threading model. And it wouldn't work in browser.

Reply:
@purplefox 

>>The generated clients have pretty ugly APIs and we'd need to wrap them in custom code to give them a nice simple idiomatic reactive streams interface.
https://grpc.io/docs/tutorials/basic/java/ I have actually seen many companies move to gRPC for bi-directional streaming recently and IMO the APIs are not that ugly. I realize this is a subjective matter. 

>>We'd also have to adapt the flow control behaviour to something reactive streams friendly - again, fiddly work that we'd have to do on both client and server side.
Using http/2 we would get some flow control, prioritization for free. Atleast, gRPC's underlying chromium stack is at the cutting edge of this. In my previous life, we have used it at massive scale to stream data back and forth between data centers and mobile phones.. 

>>We'd have no control of the threading model (esp. on the server side) - this means it couldn't use Vert.x and integrate nicely with the rest of the API using Vert.x
You may be right here.. I don't have great insights into gRPC server side threading model. 

>>Sadly HTTP2 streaming wasn't designed to present a streaming API to the client. The streaming is really as a back channel for keeping HTTP caches in sync.
I respectfully disagree here.. I think you may be talking about HTTP 1.1 SSEs, which is limiting in nature. HTTP/2 supports bi directional streams and solves many of the issues you mentioned already - multiplexing, flow control, prioritization.. Just to show how simialr they are, here is a proposal underway from Mozilla https://tools.ietf.org/html/rfc8441, that sends websocket traffic over the same http/2 connection to use the tcp connection underneath

>>This means you can't use HTTP2 as a good means of streaming to JavaScript browser clients, and this is the reason WebSockets haven't been deprecated 
Could be true. But is streaming results directly to a web browser the core use case for push queries?  I can't imagine why that would happen without a middle layer in between ksqlDB and browser.. I mean, we atleast need a load balancing or broker layer to terminate connections right.. IMO, we should not be swayed too much by the serving to browser directly use-case.. the middle layer <=> browser, can very well be using websockets. 


Nonetheless, can you clarify if we will be using thrift/protobuf, even if we build our own binary protocol? (may be I missed that detail when reading).


In summary, yes, we need a wrapper layer on top of gRPC to frame our own messages.. but there is so much already built it and a nice (sure, subjective) bi-directional API is handed back to us for building this wrapper.. And there are so many languages supported already.

Reply:
> @purplefox
> 
> > > The generated clients have pretty ugly APIs and we'd need to wrap them in custom code to give them a nice simple idiomatic reactive streams interface.
> 
> https://grpc.io/docs/tutorials/basic/java/ I have actually seen many companies move to gRPC for bi-directional streaming recently and IMO the APIs are not that ugly :) . I realize this is a subjective matter.

We need out client API to expose KSQL abstractions - the generated gRPC ones are not going to do that. Moreover we need reactive streams abstractions too, again, gRPC won't do that either, so we'd need to wrap and that is significant work.

> 
> > > We'd also have to adapt the flow control behaviour to something reactive streams friendly - again, fiddly work that we'd have to do on both client and server side.
> 
> Using http/2 we would get some flow control, prioritization for free. Atleast, gRPC's underlying chromium stack is at the cutting edge of this. In my previous life, we have used it at massive scale to stream data back and forth between data centers and mobile phones..

I'm not saying that gRPC doesn't have flow control - it does (well it's just HTTP2 flow control), what I'm saying is we'd have to adapt the flow control to the way reactive streams does flow control on both the client and server. This would be a significant amount of extra work to do on both client and server side.

> 
> > > We'd have no control of the threading model (esp. on the server side) - this means it couldn't use Vert.x and integrate nicely with the rest of the API using Vert.x
> 
> You may be right here.. I don't have great insights into gRPC server side threading model.
> 
> > > Sadly HTTP2 streaming wasn't designed to present a streaming API to the client. The streaming is really as a back channel for keeping HTTP caches in sync.
> 
> I respectfully disagree here.. I think you may be talking about HTTP 1.1 SSEs, 

No, this is HTTP2. Afaik, there is no new HTTP JavaScript API that can be used in browser. The streaming in HTTP2 was never intended as something that would be exposed to the JavaScript developer in the browser. Server push was designed as a mechanism to pre-emptively load resources in to the browsers local cache before they were requested, but from the JavaScript devs point of view nothing changes - it's an optimisation.

Now of course you can write your own client that uses HTTP2 and does expose the streaming to the end user, and that's exactly what gRPC does, but you can't use that from a browser.

This is by design, aiui, the official streaming mechanism currently is websockets and there's no plan to deprecate them and expose HTTP2 streaming in browser instead.

> which is limiting in nature. HTTP/2 supports bi directional streams and solves many of the issues you mentioned already - multiplexing, flow control, prioritization.. Just to show how simialr they are, here is a proposal underway from Mozilla https://tools.ietf.org/html/rfc8441, that sends websocket traffic over the same http/2 connection to use the tcp connection underneath

Well sure, if you write the browser you can do stuff like this, but we're talking about what APIs are available to mortal standard JavaScript devs :)

> 
> > > This means you can't use HTTP2 as a good means of streaming to JavaScript browser clients, and this is the reason WebSockets haven't been deprecated
> 
> Could be true. But is streaming results directly to a web browser the core use case for push queries? 

Probably not the most important case but certainly a useful case.

> I can't imagine why that would happen without a middle layer in between ksqlDB and browser.. I mean, we atleast need a load balancing or broker layer to terminate connections right.. 

Most loadbalancers these days are websocket friendly.

> IMO, we should not be swayed too much by the serving to browser directly use-case.. the middle layer <=> browser, can very well be using websockets.

Yes, the proposed ksqlDB client will run over websockets.

Reply:
> IMO, we should not be swayed too much by the serving to browser directly use-case.. the middle layer <=> browser, can very well be using websockets.

@vinothchandar Yeah, I agree. This is something that I haven't been explicit about when I advocated for front-end apps. Similar to how you wouldn't want a browser talking right to Postgres, you probably wouldn't want one talking straight to ksqlDB. I'm not sure how much that changes in proposed implementation, but I do think in practice there ought to be some middle tier, whether that's node.js, a web server, graphQL, or whatever.

Reply:
In most cases I think most apps will indeed have a front end and not expose ksqldb directly, however I don't think we should mandate that and leave that as a choice for the app developer. The ksqlDB client is completely agnostic to that and will work just as well in either situation. Arguably it's pretty cool to talk straight from browser to a front end of ksqldb nodes, but YMMV

Reply:
Some more thoughts on gRPC: 
I see the main value of tools like gRPC (and to a lesser extent Thrift) as great tools for quickly creating programmatic interoperability for internal service architectures. E.g, you're a big enterprise and you have lots of internal [micros]services and you want a way for your clients (internal apps and other services) to quickly use those services.
But I don't see those tools as a great fit for creating clients for public products or open source projects that we create. Why? Our client is going to be the primary way that people interact with the product - it's like the facade of your house. The form of it is going to be of fundamental importance to the user experience, therefore it's important that we control it, not someone else. If we use a tool like gRPC them we're effectively outsourcing the primary interaction point of our product to someone else. I don't think that's where we want to be.